storage_controller/src/service.rs

```diff
@@ -1014,13 +1025,37 @@ impl Service {
                             // Transition to active involves reconciling: if a node responds to a heartbeat then
                             // becomes unavailable again, we may get an error here.
                             tracing::error!(
-                                "Failed to update node {} after heartbeat round: {}",
+                                "Failed to update node state {} after heartbeat round: {}",
                                 node_id,
                                 err
                             );
                         }
                     }
                 }
+
+                // We collected all the transitions above and now we handle them.
+                let res = self.handle_node_availability_transitions(to_handle).await;
```

Comment:
These now run in a separate critical section from the `node_state_configure` transitions above. Is that safe? What if it races with another transition?

Reply:
During normal operation there's only one source of transitions, `Service::spawn_heartbeat_driver` and we always wait to apply the deltas before moving on to the next heartbeat round.
 
We also allow operators to generate transitions via `Service::external_node_configure` (hooked up to HTTP endpoint). So, yeah, in theory `Service::handle_node_availability_transitions` can race with `Service::external_node_configure`.

I'll change the locking in `Service::spawn_heartbeat_driver`.

Reply:
Done in https://github.com/neondatabase/neon/pull/9281/commits/a69de6c585d53c904c9108a924cbad569841770b

Reply:
Thanks! A secondary issue is that we'll now hold onto the locks until all transitions across all nodes have been processed, which can presumably be much longer than before when we did node by node. Can we starve out other actors? Is that a problem?

Reply:
> Can we starve out other actors? Is that a problem?

Not really. The case when we have multiple transitions from the same heartbeat round should be quite rare. All other operations that take the node locks are human operator driven. I think it's fine to wait in those scenarios.